### ISSUE TITLE
Apache return 500 error with authentication by LDAP secure port (ldaps)

### ISSUE THREAD

**Issue by maxwell-g** [smart-mobile-software/gitstack#140]
### Description

After entering ldap info into the Authentication Dialog and clicking "Test ldap Settings" button. I receive the following "Cant contact ldap server" along with the log messages that follow. (I receive a succesfull connection message if I change to a ldap protocol and a 389 port)
##### AccessLog:

GET /rest/settings/authentication/ldap/test/?protocol=ldaps&host=ldaps2.fortlewis.edu&port=636&baseDn=OU%3DUsers%2COU%3DIT%2CDC%3Dfortlewis%2CDC%3Dedu&attribute=sAMAccountName&scope=one&filter=(objectClass%3Dperson)&bindDn=CN%????????%2COU%3DSystemUsers%2CDC%3Dfortlewis%2CDC%3Dedu&bindPassword=????????????????t3&_=1350659777568 HTTP/1.1" 500 25
##### ErrorLog: (with ldap debugging turned on)

ldap_int_open_connection
ldap_connect_to_host: TCP ldaps2.fortlewis.edu:636
ldap_new_socket: 1324
ldap_prepare_socket: 1324
ldap_connect_to_host: Trying 10.44.20.10:636
ldap_pvt_connect: fd: 1324 tm: -1 async: 0
TLS trace: SSL_connect:before/connect initialization
TLS trace: SSL_connect:SSLv2/v3 write client hello A
TLS trace: SSL_connect:SSLv3 read server hello A
TLS certificate verification: depth: 1, err: 20, subject: /C=US/O=Internet2/OU=InCommon/CN=InCommon Server CA, issuer: /C=SE/O=AddTrust AB/OU=AddTrust External TTP Network/CN=AddTrust External CA Root
TLS certificate verification: Error, unable to get local issuer certificate
TLS trace: SSL3 alert write:fatal:unknown CA
TLS trace: SSL_connect:error in SSLv3 read server certificate B
TLS trace: SSL_connect:error in SSLv3 read server certificate B
TLS: can't connect: error:14090086:SSL routines:SSL3_GET_SERVER_CERTIFICATE:certificate verify failed (unable to get local issuer certificate).
ldap_err2string
### Environment

Windows 7
GitStack Demo version 2.2.4
Apache 2.2.22 bundled with GitStack
### ldap config

Protocol: ldaps
host:ldaps2.fortlewis.edu
Port:636
Base dn: OU=Users,OU=IT,DC=fortlewis,DC=edu
Attribute:sAMAccountName
Scope:one
Filter:(objectClass=person)
Bind Dn: CN=????????,OU=SystemUsers,DC=fortlewis,DC=edu
Bind password:????????

**Comment by poiuytrez** [comment]
Do you have a valid SSL certificated issues from a well known certificate authority? Which one?
Thanks.

**Comment by maxwell-g** [comment]
Yes, our ldaps cert requires a  root and intermediate certificate. They are from Certificate Authorities Comodo and Incommon.

I put all three in a ‘ldaps.cert’ file and added the following to httpd.conf:
LDAPTrustedGlobalCert CERT_BASE64 "C:\GitStack\apache\conf\ldaps.cert"
If u look at this bug in the last comment they mention apache version 22 having this problem.

https://issues.apache.org/bugzilla/show_bug.cgi?id=50630

**Comment by poiuytrez** [comment]
LDAPVerifyServerCert Off
or 
LDAPTrustedGlobalCert CERT_BASE64 "C:\GitStack\apache\conf\ldaps.cert"
seems to fix apache.

Do you have trouble to test the ldap settings on the web interface?

**Comment by maxwell-g** [comment]
Yes the problem is when I am using the test ldap settings. I have tried LDAPVerifyServerCert and LDAPTrustedGlobalCer and I receive the same message:

ldap_create
ldap_url_parse_ext(ldaps://ldaps2.fortlewis.edu:636/)
ldap_sasl_bind
ldap_send_initial_request
ldap_new_connection 1 1 0
ldap_int_open_connection
ldap_connect_to_host: TCP ldaps2.fortlewis.edu:636
ldap_new_socket: 1276
ldap_prepare_socket: 1276
ldap_connect_to_host: Trying 10.44.20.10:636
ldap_pvt_connect: fd: 1276 tm: -1 async: 0
TLS trace: SSL_connect:before/connect initialization
TLS trace: SSL_connect:SSLv2/v3 write client hello A
TLS trace: SSL_connect:SSLv3 read server hello A
TLS certificate verification: depth: 1, err: 20, subject: /C=US/O=Internet2/OU=InCommon/CN=InCommon Server CA, issuer: /C=SE/O=AddTrust AB/OU=AddTrust External TTP Network/CN=AddTrust External CA Root
TLS certificate verification: Error, unable to get local issuer certificate
TLS trace: SSL3 alert write:fatal:unknown CA
TLS trace: SSL_connect:error in SSLv3 read server certificate B
TLS trace: SSL_connect:error in SSLv3 read server certificate B
TLS: can't connect: error:14090086:SSL routines:SSL3_GET_SERVER_CERTIFICATE:certificate verify failed (unable to get local issuer certificate).
ldap_err2string

**Comment by poiuytrez** [comment]
Ok, the "test settings" and "sync" buttons triggers a python code and is not related to apache. I am pretty sure that the solution is on this page : 
http://stackoverflow.com/questions/3495739/ldap-connection-problem-with-self-signed-cert

We will fix it in the next version of GitStack.

**Comment by maxwell-g** [comment]
There were 2 ways to get around the error, by adding one of the following functions within the views.py
1.         ldap.set_option(ldap.OPT_X_TLS_REQUIRE_CERT, ldap.OPT_X_TLS_NEVER)
2.   ldap.set_option(ldap.OPT_X_TLS_CACERTFILE,"C:\opends-cert.pem")

I now sync and save my ldap users. I am also able to select and assign a ldap user to a repository.

I however run into problems when I select the repository and try to login using these ldap users. I enter the user name and password and it returns with user name not found.
I think the problem is within the Authentication.class.php. Within the authenticate function, it successfully reads the repository users but  $_SERVER['PHP_AUTH_USER'] is not defined.

Is there a work around for this?
Gary Maxwell

**Comment by maxwell-g** [comment]
Sorry I was wrong on the previous email. As stated my ldaps test button does work after implementing 1 of these 2 changes in views.py:
1. ```
     ldap.set_option(ldap.OPT_X_TLS_REQUIRE_CERT, ldap.OPT_X_TLS_NEVER)
   ```
   1.   ldap.set_option(ldap.OPT_X_TLS_CACERTFILE,"C:\opends-cert.pem")

I am now getting a similar message when I try to login using one of these ldap users into a specific repository. The checkLdapUser() function of Authentication.class.php produces the following message: after the bind has occurred:
Error Binding to LDAP: error:14090086:SSL routines:SSL3_GET_SERVER_CERTIFICATE:certificate verify failed

Both the Python and PHP code seems to be giving the same type of message about validating against the cert. Is it possible that I can have a problem on Python and PHP? Is there a way to turn off the cert verification in the PHP code?

From: Maxwell, Gary
Sent: Tuesday, October 23, 2012 3:19 PM
To: 'smart-mobile-software/gitstack'; smart-mobile-software/gitstack
Subject: RE: [gitstack] Apache return 500 error with authentication by LDAP secure port (ldaps) (#140)

There were 2 ways to get around the error, by adding one of the following functions within the views.py
1. ```
       ldap.set_option(ldap.OPT_X_TLS_REQUIRE_CERT, ldap.OPT_X_TLS_NEVER)
   ```
   1.   ldap.set_option(ldap.OPT_X_TLS_CACERTFILE,"C:\opends-cert.pem")

I now sync and save my ldap users. I am also able to select and assign a ldap user to a repository.

I however run into problems when I select the repository and try to login using these ldap users. I enter the user name and password and it returns with user name not found.
I think the problem is within the Authentication.class.php. Within the authenticate function, it successfully reads the repository users but  $_SERVER['PHP_AUTH_USER'] is not defined.

Is there a work around for this?
Gary Maxwell

**Comment by poiuytrez** [comment]
That's good news.
Yes there is a workaround for it : https://github.com/smart-mobile-software/gitstack/issues/138
The patch will be integrated in GitStack 2.2.5.

**Comment by maxwell-g** [comment]
Is there a scheduled time when 2.2.5 will be released? We are under a time crunch.

**Comment by poiuytrez** [comment]
GitStack 2.2.5 will be released in approximately 2 weeks. You can modify
the code yourself during the mean time using the suggested workarounds.

**Comment by poiuytrez** [comment]
Fixed in #138.
